### PR TITLE
SQL: Bubble up negations in expressions

### PR DESCRIPTION
**Pull request by bpintea**
This PR pushes the negations in expressions up the AST:
 * MUL/DIV - NEG - x => NEG - MUL/DIV - x
 * NEG - MUL/DIV - NEG - x => MUL/DIV - x
 * NEG - NEG - x => x

This can potentially simplify the expression, but canceling out two
negations, but also prepares the expression to later be simplified, if
part of a binary comparison.

### CODE REVIEW

**Comment by bpintea** [comment]
This is a simple extraction from #66022.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java [hunk 1]
```diff
@@ -1137,6 +1142,51 @@ protected Expression rule(Expression e) {
         }
     }
 
+    public static final class BubbleUpNegations extends OptimizerExpressionRule {
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1145]: Kept the `BubbleUpNegations` naming, but not strongly attached to it -- will rename if anyone feels significantly better about `SimplifyNegations`.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java [hunk 2]
```diff
@@ -1137,6 +1142,51 @@ protected Expression rule(Expression e) {
...
+                    if (operation.right() instanceof Neg) {
+                        Neg neg = (Neg) operation.right();
+                        return new Neg(operation.source(), operation.replaceChildren(List.of(operation.left(), neg.field())));
+                    }
+                }
+            }
+            return e;
+        }
+
+        public static boolean isMulOrDiv(ArithmeticOperation operation) {
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1184]: `public` as extracted from #66022, where it's being reused in another rule.
Will make eventually `protected`, however.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java (hunk 1 above)
Comment (palesz) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1145]: We left out the negations from the discussed operators. Sorry for this late discovery, but just realized that the `Integer.{MIN,MAX}_VALUE` differences would bite us (small chance, but still):

```
-- with the negations in place, this works
SELECT -1073741824 * 2 / -3

-- after optimizing the negations this blows up with integer overflow
SELECT 1073741824 * 2 / 3
```

**Review by palesz — CHANGES_REQUESTED** [review]
One main concern that can cause `ArithmeticException` on queries that work without the optimizer in place.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java [hunk 3]
```diff
@@ -1137,6 +1142,51 @@ protected Expression rule(Expression e) {
...
+                if (child instanceof ArithmeticOperation) { // cancels out NEG - MUL/DIV - NEG - x ==> MUL/DIV - x
+                    ArithmeticOperation operation = (ArithmeticOperation) child;
+                    if (isMulOrDiv(operation)) {
+                        if (operation.left() instanceof Neg) {
+                            return operation.replaceChildren(List.of(((Neg) operation.left()).field(), operation.right()));
+                        }
+                        if (operation.right() instanceof Neg) {
+                            return operation.replaceChildren(List.of(operation.left(), ((Neg) operation.right()).field()));
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1164]: Coupled with the binary comp optimisation executing subsequently and reducing the negation, the last rule was superfluous, but the second was necessary (and is also faster converging).
As a standalone optimisation however, the 2nd rule is indeed redundant.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java (hunk 1 above)
Comment (bpintea) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1145]: Correct, thanks for pointing it out.

I guess a more immediate example is simply `-int * 2`, the before/after difference being the working `NEG(int) * 2` vs the failing `NEG(int * 2)` for the above "-MIN_VALUE/2" value.

**Comment by bpintea** [comment]
Closing as behaviour changing optimisation.

**Pull request closed without merging**